### DISCUSSION TITLE
GB10 (Blackwell SM121) — 194 experiments, SDPA fallback reshapes compute-optimal point

### DISCUSSION THREAD

**Discussion by cerebustech-dev** [karpathy/autoresearch#108] [Show and tell]
*This is an AI-generated summary from an autoresearch agent (Claude) running on my behalf on an NVIDIA GB10.*

## Hardware

**NVIDIA GB10 (DGX Spark)** — Grace Blackwell, SM121, 128GB unified memory, CUDA 13.1, aarch64 Linux.

This is the first autoresearch run on Blackwell silicon. The key constraint: **Flash Attention 3 has no kernel for SM121**, so we fall back to `torch.nn.functional.scaled_dot_product_attention` (SDPA). This makes attention ~30x slower than H100, which fundamentally reshapes what the optimal model looks like.

## Setup

Used [superoo7/autoresearch-cu130](https://github.com/superoo7/autoresearch-cu130) fork (PR #30 on upstream) which adds the SDPA fallback. PyTorch cu128 wheels work on aarch64 with CUDA 13.1 (forward compatible). Triton needs `TRITON_PTXAS_PATH` pointed at system ptxas (bundled ptxas lacks SM121a support).

Before starting the autonomous agent, we did a manual hyperparameter sweep to find a reasonable GB10 baseline, since the stock depth=8 config only manages 31 steps in 5 minutes (vs ~950 on H100). We found **depth=2, ASPECT_RATIO=128, MATRIX_LR=0.12** as the starting point (val_bpb=1.379).

The agent (Claude, `--dangerously-skip-permissions`) then ran autonomously for **~26 hours**, completing **194 experiments** (36 keeps, 155 discards, 3 crashes).

### All kept experiments (chronological)

```
commit   val_bpb   mem_gb  description
677c9c3  1.379414   9.1    baseline (depth=2, AR=128, matrix_lr=0.12)
5182623  1.355546   9.1    switch window pattern SSSL->L (all full context)
407b2b5  1.333809  18.1    DEVICE_BATCH_SIZE 128->256 (no grad accum)
68d6411  1.325770  18.1    replace ReLU^2 MLP with SwiGLU
6379777  1.324470  18.4    VE on all layers (both layers)
257f563  1.287860   9.3    halve batch (TOTAL=2^18 DEVICE=128) - 305 steps
7085bb8  1.270482   4.7    halve batch again (TOTAL=2^17 DEVICE=64) - 598 steps
7e8ff16  1.268107   2.4    halve batch again (TOTAL=2^16 DEVICE=32) - 1180 steps
8f77501  1.263425   2.4    MATRIX_LR 0.12->0.08
e8663d1  1.263002   2.4    MATRIX_LR 0.08->0.06
67b9dc4  1.261371   2.4    EMBEDDING_LR 0.6->0.3
32904d3  1.236567   4.1    depth 2->3 (small batch allows it!)
35242f1  1.236276   4.1    WEIGHT_DECAY 0.2->0.1
496e275  1.234808   4.1    FINAL_LR_FRAC 0.0->0.1
8bb0546  1.234587   4.1    WARMDOWN_RATIO 0.5->0.6
376de86  1.229336   4.1    SCALAR_LR 0.5->0.2
87b6a3d  1.228481   4.1    torch.compile max-autotune
0d27394  1.228189   4.1    Muon beta2 0.95->0.90
7a9b193  1.228132   4.1    WARMDOWN_RATIO 0.6->0.7
f9beff8  1.227673   4.1    Muon momentum warmup 300->500 steps
aed4fde  1.226656   4.1    Muon momentum warmup 500->700 steps
c23e2b1  1.225715   4.1    Muon momentum warmup 700->1000 steps
35d51f4  1.224658   4.1    Muon momentum warmup 1000->2000 steps
343eeaf  1.224278   4.1    constant Muon momentum 0.85
e66577e  1.222581   4.1    WD=0.01 on value embeddings
5ee6d23  1.222414   4.1    momentum warmdown 0.85->0.80
d9c3589  1.221591   4.1    softcap 15->12
0214405  1.221542   4.1    Muon ns_steps 5->6
0d5205a  1.220835   4.1    VE gate channels 32->16
f611438  1.219994   4.1    WD constant then decay in warmdown
0b491a3  1.217926   4.1    token shift (half channels shifted by 1 pos before QKV)
0b2618c  1.216209   4.1    token shift 1/4 channels (better than 1/2)
5d5f0a8  1.215238   4.1    token shift 1/8 channels (better than 1/4)
568fc4b  1.215065   4.1    softcap 12->11
d025feb  1.214183   4.2    token shift on K only (Q and V unshifted)
dbcd0e0  1.213946   4.2    WARMDOWN_RATIO 0.7->0.75
```

### Notable crashes/failures

- **Weight tying**: 3.216 BPB — catastrophic (matches #43's finding)
- **Parallel attn+MLP**: 1.249 — much worse (matches #43)
- **Label smoothing 0.1**: 1.543 — incompatible with BPB evaluation
- **Depthwise Conv1d**: crashed Muon optimizer (1D params not supported)
- **QK bias**: also crashed Muon

### 1. Compute-optimal architecture is radically different

On H100, the sweet spot is depth 9 with 512-dim (~50M params, ~950 steps). On GB10 with SDPA fallback, it's **depth 3 with 384-dim** (~10M params, ~1180 steps). The attention bottleneck means you want a much smaller model that iterates faster.

### 2. Batch halving is the biggest single win (again)

The agent tripled down on this — halving batch from 2^19 → 2^18 → 2^17 → 2^16, each time getting more steps and better val_bpb. This matches #43 and #32's finding that smaller batches dominate. On GB10 this was even more dramatic because more steps = more value when each step is expensive.

### 3. Cross-platform validation of key findings

Confirmed across H100 (#32, #43) and GB10:
- **Batch halving**: biggest win on all platforms
- **Warmdown 0.5→0.7+**: consistent improvement (we landed at 0.75)
- **Warmup is harmful**: tested twice, hurt both times
- **Weight tying catastrophic**: broken on both platforms
- **Parallel attn+MLP**: worse on both platforms
- **VE weight decay**: small but real improvement (flagged as "needs validation" in #66 — here's your validation)

### 4. SwiGLU > ReLU^2

The agent replaced the ReLU^2 MLP with SwiGLU (gated linear unit with SiLU activation). This was a clean win with no speed penalty at the same parameter count.

### 5. Token shifting on K only

The most creative finding. The agent discovered that shifting 1/8 of channels by 1 position **on keys only** (not queries or values) gives the model cheap local context awareness. It systematically tested:
- 1/2 channels: 1.218 (too much)
- 1/4 channels: 1.216
- **1/8 channels: 1.215** (sweet spot)
- 1/16 channels: 1.224 (too little)
- K-only vs Q+K: K-only wins (1.214 vs 1.219)

The intuition: keys benefit from knowing what the previous token was, but queries and values don't need this.

### 6. torch.compile helps even on GB10

Despite `"Not enough SMs to use max_autotune_gemm mode"` warning, `torch.compile(mode="max-autotune")` still improved results. The ~100s compile warmup is worth it because the fused optimizer kernels save ~30% per step.

## Final optimized config

```python
# Architecture
DEPTH = 3
ASPECT_RATIO = 128    # -> 384-dim, 3 heads
HEAD_DIM = 128
WINDOW_PATTERN = "L"  # all full-context (avoids slow attn_mask path)

# Optimization
TOTAL_BATCH_SIZE = 2**16  # 65K tokens (small batch = more steps)
DEVICE_BATCH_SIZE = 32
MATRIX_LR = 0.06
EMBEDDING_LR = 0.3
SCALAR_LR = 0.2
WEIGHT_DECAY = 0.1
WARMDOWN_RATIO = 0.75
FINAL_LR_FRAC = 0.1

# Architectural changes from stock:
# - SwiGLU MLP (replaces ReLU^2)
# - VE on all layers
# - VE gate channels 16 (was 32)
# - VE weight decay 0.01
# - Token shift 1/8 channels on K only
# - Softcap 11 (was 15)
# - Muon: constant momentum 0.85 with warmdown to 0.75, beta2=0.90, ns_steps=6
# - WD constant then decay in warmdown
```

## Comparison with discussion #43 (H100)

| | H100 (#43) | GB10 (this run) |
|---|---|---|
| **Experiments** | 126 | 194 |
| **Duration** | 10.5 hours | ~26 hours |
| **Starting val_bpb** | 0.998 | 1.379 |
| **Final val_bpb** | 0.970 | 1.214 |
| **Improvement** | 0.028 | 0.165 |
| **Optimal depth** | 9 | 3 |
| **Optimal batch** | 262K | 65K |
| **Steps/experiment** | ~950 | ~1180 |
| **Peak VRAM** | ~44 GB | ~4 GB |

The val_bpb numbers aren't directly comparable (different step counts and throughput), but the *direction* of findings is remarkably consistent across platforms.

## Thoughts

This validates Karpathy's thesis that autoresearch finds the optimal model for YOUR hardware. The GB10 and H100 arrive at completely different architectures but through similar optimization strategies. The fact that an autonomous agent independently rediscovered batch halving, warmdown tuning, and VE weight decay — findings from separate H100 runs — is a nice confirmation that these are robust results, not noise.

The SDPA bottleneck on GB10 is the dominant constraint. If/when Flash Attention gets SM121 support, the compute-optimal point will shift dramatically. Would be interesting to re-run then and see how the architecture changes.

Full `results.tsv` with all 194 experiments is available in the branch `autoresearch/mar8-gb10`.

**Comment by aniruddhaadak80** [comment]
Wow, running this on a GB10 is super cool. The SDPA fallback altering the optimal setup makes a lot of sense since FA3 is so highly tuned for raw Ampere/Hopper throughput. By missing out on those specialized tensor core usages, SDPA probably shifts the memory-bound vs compute-bound curve quite a bit.

I’m curious how much of a performance gap you saw between the native kernels and the fallback here. Have you noticed any unusual behavior with the model's convergence rates adjusting to this? Super interesting data point to have.

**Comment by cerebustech-dev** [comment]
Thanks! To answer your questions:

**Performance gap:** The SDPA fallback runs at ~15s/step vs H100's ~0.3s/step with FA3 — roughly 50x slower on the attention operation itself. The interesting thing is it's not just raw throughput loss. The sliding window path in SDPA requires materializing a full T×T boolean mask (`attn_mask`), while the `is_causal=True` path avoids this. One of the agent's first wins was switching to all-L (full context) window pattern specifically to stay on the faster `is_causal` path — a choice that wouldn't matter at all on H100 with FA3.

**Convergence:** No unusual behavior — the model converges normally, it just sees fewer tokens per unit time. The agent compensated by finding a much smaller model (depth 3, 384-dim, ~10M params) that iterates faster. At the optimal batch size (2^16), it actually achieves ~1180 steps per 5-minute run — *more* steps than the H100 baseline (~950), just with 8x smaller batches. So the convergence rate per step is fine; the hardware constraint reshapes *what* you train, not *how well* it trains.

**Update: still running!** The agent is now at **316 experiments over 2 days** and has pushed val_bpb from 1.214 to **1.210**. Notable new finds since the original post:

- **Post-norm on attention output** (RMS norm before c_proj): 1.212 — an architectural change, not just hyperparameter tuning
- **Embedding noise injection** (0.008 std Gaussian noise on embeddings during training): 1.212 — a regularization trick that helps with the small batch regime
- It also tried **MoE (mixture of experts)** which crashed Muon, and **multi-token prediction** which was too slow. Bold ideas, even if they didn't land.

The agent is deep in diminishing returns now (68 experiments for 1 keep since last update) but still faithfully following program.md's "NEVER STOP" instruction.

**Comment by fayerman-source** [comment]
Follow-up data point from an RTX 5070 12 GB fork run.

I tested the main SDPA-specific ideas from this thread in a fixed 5-run sequence:
- baseline (`SSSL`): `val_bpb 1.269448`, ~79k tok/s
- `WINDOW_PATTERN=L`: `1.189742`, ~105k tok/s
- `+ WARMDOWN_RATIO=0.75`: `1.187303`, ~105k tok/s
- `+ FINAL_LR_FRAC=0.05`: `1.183502`, ~106k tok/s
- `+ small AdamW WD on embeddings/value embeds/lm_head`: `1.184761`, ~103k tok/s

So on 5070/SDPA, `WINDOW_PATTERN=L` reproduced strongly, warmdown `0.75` helped a bit more, and a small LR floor helped again. The targeted AdamW decay mix did not beat the no-extra-WD run here.

I also upstreamed the code side separately: PR #161 adds a non-Hopper SDPA fallback plus cached sliding-window masks: https://github.com/karpathy/autoresearch/pull/161

**Comment by heidiEC** [comment]
Nice! I added this to the graph prior on Discussion #137. You can also query the graph directly (instructions on the ticket) and i'm building an interface to viz.

**Comment by Nyrok** [comment]
The "NEVER STOP" detail from program.md is a good example of why explicit instructions beat implicit ones. The agent follows it across 316 experiments because it is stated as a clear constraint. If that same rule were buried in a paragraph about being thorough, the model would treat it as a preference rather than a directive.

Same principle applies when writing the initial task definition: goal, stopping criteria, and constraints as distinct labeled sections tend to produce more predictable behavior than a single prose block.

I've been building [flompt](https://flompt.dev) for exactly that, a visual builder that decomposes task instructions into 12 typed semantic blocks and compiles to Claude-optimized XML. Open-source: github.com/Nyrok/flompt
